[BD-5] [TNL-7282] [BB-2621] Support flexible peer grading

lullis opened the pull request:
This PR introduces support for flexible grading, i.e, a student can
receive their score to an ORA submission before getting the minimum
required peer grades, if at least floor(70% of defined minimum peer
grades are given)

**JIRA tickets**:
 - https://tasks.opencraft.com/browse/BB-2621
 - https://openedx.atlassian.net/browse/TNL-7282

Screenshots: ![Added "Enable Peer Flexible Grading" setting](https://user-images.githubusercontent.com/16288/89134994-40291b80-d52a-11ea-8dc4-6061378de950.png)

**What changed?**

 - Added option to peer assessment grading options to enable/disable flexible grading
 - Added `enable_flexible_grading` attribute to peer assessment mixin
 - Added `_required_peer_grades` function to `openassessment/assessment/api/peer.py` determine if student can get submission score

**Reviewers**
- [ ] @xitij2000

Commit:
Add support to flexible grading.

xitij2000 commented:
@lullis Could you rebase this PR on the current master? It's currently including a lot of other commits making it hard to review.

xitij2000 commented on openassessment/assessment/api/peer.py:
```diff
@@ -168,7 +179,7 @@ def get_score(submission_uuid, peer_requirements):
     ).order_by('-assessment')
 
-    submission_finished = items.count() >= peer_requirements["must_be_graded_by"]
+    submission_finished = items.count() >= _required_peer_grades(workflow, peer_requirements)
```
I think it might make sense to move this entire evaluation into the function, i.e. call the function: `_check_submission_finished` or `_is_submission_finished` etc..

xitij2000 commented on openassessment/templates/openassessmentblock/edit/oa_edit_peer_assessment.html:
```diff
@@ -26,13 +26,29 @@
                 <li class="field comp-setting-entry">
                     <div class="wrapper-comp-setting">
-                        <label for="peer_assessment_graded_by" class="setting-label"> {% trans "Graded By" %}</label>
+                        <label for="peer_assessment_graded_by" class="setting-label">{% trans "Graded By" %}</label>
```
nit: avoid unnecessary diff.

xitij2000 commented on openassessment/templates/openassessmentblock/edit/oa_edit_peer_assessment.html:
```diff
@@ -26,13 +26,29 @@
                 </li>
                 <li class="field comp-setting-entry">
-                    <div class="wrapper-comp-setting">
+                  <div class="wrapper-comp-setting">
+                    <label for="peer_assessment_enable_flexible_grading" class="setting-label">
+                      {% trans "Enable Flexible Peer Grade Averaging"%}
+                    </label>
+                    <select id="peer_assessment_enable_flexible_grading" class="input setting-input">
+                      <option value="0">{% trans "False"%}</option>
+                      <option value="1" {% if assessments.peer_assessment.enable_flexible_grading %}selected="true"{% endif %}>
+                        {% trans "True"%}
+                      </option>
+                    </select>
+                  </div>
+                  <p class="setting-help">
+                    {% trans "Learners who have received a minimum of {floor-30% of required peer-reviews configuration value} who have waited longer than 7 days will be given their grade instead of having to wait longer for more peer reviews"%}
+                  </p>
+                </li>
+                <li class="field comp-setting-entry">
+                  <div class="wrapper-comp-setting">
```
nit: doesn't seem to be matching the code style of the rest of the file.

xitij2000 commented on openassessment/xblock/schema.py:
```diff
@@ -121,6 +121,7 @@ def datetime_validator(value):
             'required': bool,
             'must_grade': All(int, Range(min=0)),
             'must_be_graded_by': All(int, Range(min=0)),
+            Required('enable_flexible_grading', default=True): bool,
```
It should default to false, since that is the current behaviour.

lullis replied:
I'd agree with you and keep the existing behavior, but the ticket description says "By default for problems we’d like to have this be enabled but for existing problems this might be disabled.", which led me to switch to true.

xitij2000 replied:
Alright, in that case, we need to make sure that any existing blocks that don't have this field are handled properly, i.e. if this field is undefined, it should default to false. 

So I think it can default to false in the schema, but default to true for the data used to fill in the default values in `defaults.py`.

xitij2000 commented on openassessment/xblock/static/js/src/studio/oa_edit_assessment.js:
```diff
@@ -57,6 +57,7 @@ OpenAssessment.EditPeerAssessmentView.prototype = {
      {
          must_grade: 5,
          must_be_graded_by: 2,
+         enable_flexible_grading: true,
```
nit: default to false.

xitij2000 reviewed (commented):
This is just a first pass of comments. I will test properly and review later.

Commit:
Move function to calculate required peer grades to check_submission

Commit:
Reformat HTML template

Commit:
Change Peer Assessment schema default to false

For backwards compatibility, the value of enable_flexible_grading
should be false

Commit:
Refactor to share flexible grading logic in multiple place.

Commit:
Fix Spacing after template tags

Commit:
Parse & export enable_flexible_grading in xml

Commit:
tz templatetags required for utc filter

Commit:
Existing tests / data might not have enable_flexible_grading

Commit:
Tests should expect enable_flexible_grading exists

Commit:
Existing tests should expect enable_flexible_grading to be false

Commit:
Fix merge (#1478)

* fix template merge conflict gunk

* fix bad merge

* v2.9.6

* test docstring

Commit:
Test if flexible grading works

Commit:
Fix quality issue

Commit:
Expect enable_flexible_grading to exists

Commit:
update translations

Commit:
fix(i18n): update translations

Commit:
Remove unwanted fields

Commit:
Use Submission.submitted_at for calculate elapsed date. Rearrange the flow.

Commit:
Js build

Commit:
js package.json updates

Commit:
translation changes

Commit:
don't show leaderboard for teams oras, also test ui_models (#1479)

* don't show leaderboard for teams oras, also test ui_models

* fix quality complaints about tests

* fix typo

* 2.9.7 version bump

* template updates for setttings

* make javascript sass

* translations

* banner css update

* update translations for added period

* one more banner css update

* eslint fix

* Update trannslations

Co-authored-by: Nathan Sprenkle <nsprenkle@edx.org>

nsprenkle committed:
Include team fields in ORA XML import/export (#1482)

Commit:
Fix code style issues & suggestion to move magic numbers to const

Commit:
quality fix - spacing

Commit:
bump up the version

Commit:
use ddt instead of array

Commit:
text changed

Commit:
translation updated
